File: Tests/AINotebookCoreTests/EmbedderTests.swift
import XCTest
@testable import AINotebookCore

@MainActor
final class EmbedderTests: XCTestCase {

    final class MockEmbeddingClient: EmbeddingProducing, @unchecked Sendable {
        var calls: [[String]] = []
        var dim: Int = 4
        func embed(model: String, inputs: [String]) async throws -> [[Float]] {
            calls.append(inputs)
            return inputs.map { _ in (0..<self.dim).map { _ in Float.random(in: -1...1) } }
        }
    }

    func testEmbedAllInsertsRowsForEveryChunk() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: (0..<5).map { ChunkDraft(text: "c\($0)", tokenCount: 1) }
        )
        let client = MockEmbeddingClient()
        let embedder = Embedder(store: store, client: client, model: "m", batchSize: 2)
        let count = try await embedder.embedAllPending()
        XCTAssertEqual(count, 5)
        XCTAssertEqual(try store.unembeddedCount(model: "m"), 0)
        XCTAssertEqual(client.calls.count, 3, "should batch 2+2+1")
        XCTAssertEqual(client.calls.map(\.count), [2, 2, 1])
    }

    func testEmbedAllSkipsAlreadyEmbedded() async throws {
        let store = try NotebookStore(path: .inMemory)
        let nb = try store.createNotebook(name: "NB")
        let s = try store.createSource(
            notebookId: nb.id!, type: .text, title: "x", uri: nil, rawPath: nil
        )
        try store.replaceChunks(
            sourceId: s.id!,
            chunks: [
                ChunkDraft(text: "a", tokenCount: 1),
                ChunkDraft(text: "b", tokenCount: 1)
            ]
        )
        let chunks = try store.chunks(sourceId: s.id!)
        try store.storeEmbedding(
            chunkId: chunks[0].id!, model: "m",
            vector: EmbeddingVector(values: [1, 0, 0, 0])
        )

        let client = MockEmbeddingClient()
        let embedder = Embedder(store: store, client: client, model: "m", batchSize: 10)
        let count = try await embedder.embedAllPending()
        XCTAssertEqual(count, 1)
        XCTAssertEqual(client.calls.count, 1)
        XCTAssertEqual(client.calls[0], ["b"])
    }
}

File: windows/tests/AINotebook.Core.Tests/Rag/EmbedderTests.cs
using AINotebook.Core;
using AINotebook.Core.Models;
using AINotebook.Core.Rag;
using AINotebook.Core.Storage;
using AINotebook.Core.Tests.Helpers;
using Xunit;

namespace AINotebook.Core.Tests.Rag;

public class EmbedderTests
{
    // EmbedderTests.testEmbedAllInsertsRowsForEveryChunk
    [Fact]
    public async Task EmbedAllInsertsRowsForEveryChunkInBatches()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var src = store.CreateSource(nb.Id!.Value, SourceType.Text, "S", null, null);
        store.ReplaceChunks(src.Id!.Value, new[]
        {
            new ChunkDraft("c0", 1, null), new ChunkDraft("c1", 1, null),
            new ChunkDraft("c2", 1, null), new ChunkDraft("c3", 1, null),
            new ChunkDraft("c4", 1, null),
        });

        var client = new MockEmbeddingClient();
        var embedder = new Embedder(store, client, "m", batchSize: 2);
        var written = await embedder.EmbedAllPendingAsync();

        Assert.Equal(5, written);
        Assert.Equal(0, store.UnembeddedChunks("m", 100).Count);
        Assert.Equal(3, client.Calls.Count);
        Assert.Equal(new[] { 2, 2, 1 }, client.Calls.Select(c => c.Length).ToArray());
    }

    // EmbedderTests.testEmbedAllSkipsAlreadyEmbedded
    [Fact]
    public async Task EmbedAllSkipsAlreadyEmbedded()
    {
        var store = new NotebookStore(StorePath.InMemory);
        var nb = store.CreateNotebook("N", "");
        var src = store.CreateSource(nb.Id!.Value, SourceType.Text, "S", null, null);
        store.ReplaceChunks(src.Id!.Value, new[]
        {
            new ChunkDraft("a", 1, null),
            new ChunkDraft("b", 1, null),
        });
        // Pre-embed the first chunk ('a') for model 'm'.
        var first = store.UnembeddedChunks("m", 1)[0];
        store.StoreEmbedding(first.Id!.Value, "m", new EmbeddingVector(new[] { 0.1f, 0.2f, 0.3f, 0.4f }));

        var client = new MockEmbeddingClient();
        var embedder = new Embedder(store, client, "m", batchSize: 10);
        var written = await embedder.EmbedAllPendingAsync();

        Assert.Equal(1, written);
        Assert.Single(client.Calls);
        Assert.Equal(new[] { "b" }, client.Calls[0]);
    }
}
